Add message check support to StreamServerConnection

szewai opened the pull request:
#### 773d3731d40eda83458bd4224800cf0ba4f3949f
<pre>
Add message check support to StreamServerConnection
<a href="https://bugs.webkit.org/show_bug.cgi?id=283399">https://bugs.webkit.org/show_bug.cgi?id=283399</a>
<a href="https://rdar.apple.com/140257762">rdar://140257762</a>

Reviewed by Kimmo Kinnunen.

MESSAGE_CHECK currently invokes markCurrentlyDispatchedMessageAsInvalid() on connection when check fails. To enable
message check on StreamServerConnection, add StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid(). With
this change, RemoteGraphicsContextGL could do message check in message handler functions and we could remove EnabledIf
annotation in RemoteGraphicsContextGL.messages.in. EnabledIf currently overlaps with both EnabledBy and MESSAGE_CHECK,
and we will drop it soon to avoid confusion.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::messageCheck):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createExternalImage):
(deleteExternalImage):
(bindExternalImage):
(createExternalSync):
(deleteExternalSync):
(enableRequiredWebXRExtensions):
(addFoveation):
(enableFoveation):
(disableFoveation):
(framebufferDiscard):
(framebufferResolveRenderbuffer):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Tools/Scripts/generate-gpup-webgl:
(cpp_func):
(cpp_func.__init__):
(webkit_ipc_msg):
(webkit_ipc_msg.__init__):
(webkit_ipc_msg.__str__):
(webkit_ipc_cpp_impl.__init__):
(create_cpp_arg_list):
(create_cpp_func):
(main):

Canonical link: <a href="https://commits.webkit.org/286924@main">https://commits.webkit.org/286924@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5970e1d19d626ee006c8cce0c7f05c18691d98a2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77519 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30435 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82109 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28801 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79636 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65704 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4851 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60743 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18741 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80586 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50714 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41017 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48115 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24031 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27125 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69225 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24369 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83509 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4899 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68973 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5055 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66502 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68238 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17056 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12259 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10354 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4846 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/4865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/8300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6624 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Tools/Scripts/generate-gpup-webgl:
```diff
@@ -879,6 +878,9 @@ class webkit_ipc_cpp_impl(object):
         self.pre_call_stmts += ["assertIsCurrent(workQueue());"]
+        if self.message_check:
+            self.pre_call_stmts += [f"messageCheck({self.message_check});"]
+        # self.pre_call_stmts += [f"messageCheck(\"{self.name}\"_s);"]
```
Remove commented code here.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -454,6 +456,13 @@ Ref<RemoteVideoFrameObjectHeap> RemoteGraphicsContextGL::protectedVideoFrameObje
 }
 #endif
 
+void RemoteGraphicsContextGL::messageCheck(bool assertion)
```
Yeah, this is unfortunate. I should probably improve the the generator, move the code to a generated .cpp file..

szewai committed:
Add message check support to StreamServerConnection
https://bugs.webkit.org/show_bug.cgi?id=283399
rdar://140257762

Reviewed by Kimmo Kinnunen.

MESSAGE_CHECK currently invokes markCurrentlyDispatchedMessageAsInvalid() on connection when check fails. To enable
message check on StreamServerConnection, add StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid(). With
this change, RemoteGraphicsContextGL could do message check in message handler functions and we could remove EnabledIf
annotation in RemoteGraphicsContextGL.messages.in. EnabledIf currently overlaps with both EnabledBy and MESSAGE_CHECK,
and we will drop it soon to avoid confusion.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::messageCheck):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createExternalImage):
(deleteExternalImage):
(bindExternalImage):
(createExternalSync):
(deleteExternalSync):
(enableRequiredWebXRExtensions):
(addFoveation):
(enableFoveation):
(disableFoveation):
(framebufferDiscard):
(framebufferResolveRenderbuffer):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::markCurrentlyDispatchedMessageAsInvalid):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Tools/Scripts/generate-gpup-webgl:
(cpp_func):
(cpp_func.__init__):
(webkit_ipc_msg):
(webkit_ipc_msg.__init__):
(webkit_ipc_msg.__str__):
(webkit_ipc_cpp_impl.__init__):
(create_cpp_arg_list):
(create_cpp_func):
(main):

Canonical link: https://commits.webkit.org/286924@main
